### DISCUSSION TITLE
A text decompressed and compressed again with a different results

### DISCUSSION THREAD

**Discussion by ghost** [icsharpcode/SharpZipLib#800] [Q&A]
Question:
step 1: decompress a text named A
step 2: compressed A to new text B
step 3  the content of A is not same with B, why?

Environment:
SharpZipLib version="1.3.3" targetFramework="net452"

<details>
<summary>Sourcecode of class Program</summary>

```cs
public static Bitmap ArrayToBitmap(byte[] bytes, int width, int height, System.Drawing.Imaging.PixelFormat pixelFormat)
{
    var image = new Bitmap(width, height, pixelFormat);
    var imageData = image.LockBits(new System.Drawing.Rectangle(0, 0, image.Width, image.Height),
                ImageLockMode.ReadWrite, pixelFormat);
    try
    {
        Marshal.Copy(bytes, 0, imageData.Scan0, bytes.Length);
    }
    finally
    {
        image.UnlockBits(imageData);
    }
    return image;
}

public static byte[] Decompress(byte[] data)
{
    // Create the compressor with highest level of compression
    Inflater inflater = new Inflater(true);
    inflater.SetInput(data);

    // Compress the data
    byte[] buf = new byte[1024];
    int count = 0;
    using (MemoryStream memory = new MemoryStream())
    {
        while (!inflater.IsFinished)
        {
            count = inflater.Inflate(buf);
            if (count > 0)
            {
                memory.Write(buf, 0, count);
            }
            else
            {
                break;
            }
        }

        count = inflater.Inflate(buf);
        if (count > 0)
        {
            memory.Write(buf, 0, count);
        }

        return memory.ToArray();
    }
}

public static byte[] Compress(byte[] data)
{
    // Create the compressor with highest level of compression
    Deflater deflater = new Deflater(Deflater.DEFAULT_COMPRESSION, true);
    deflater.SetStrategy(DeflateStrategy.Default);
    deflater.SetInput(data);
    deflater.Finish();

    // Compress the data
    byte[] buf = new byte[1024];
    int count = 0;
    using (MemoryStream memory = new MemoryStream())
    {
        while (!deflater.IsFinished)
        {
            count = deflater.Deflate(buf);
            if (count > 0)
            {
                memory.Write(buf, 0, count);
            }
            else
            {
                break;
            }
        }

        count = deflater.Deflate(buf);
        if (count > 0)
        {
            memory.Write(buf, 0, count);
        }

        return memory.ToArray();
    }
}

static void Main(string[] args)
{
    //raw data
    string zipString = "eJztlDEOgzAMRY0YsqDmBvVRci02OBpH4QiMDBHpVj+QojaVEFDxp6fIdr4dKyJfy/UHsKaUSvgon1vhvMrEe7CCO5SJhvWMMiNSB+OAMotFpAieMoyrkLuuybvogd7oedULe2TvnAlnxRl+VPGeVKM7FcvkydZZDMYdHiaANbMbv6t0nk9p3f+wkHdVUxsrtkH98OYAVmf8yP4+t66iF1UlgLg=";
    var a = Convert.FromBase64String(zipString).ToArray();
    var z64 = a.Skip(2).ToArray();//skip the header "eJ"
    byte[] uzipZ64 = Decompress(z64);
    int bytesPerRow = 12;
    int width = bytesPerRow * 8;
    int height = uzipZ64.Length / bytesPerRow;
    var decodedBitmap = ArrayToBitmap(uzipZ64, width, height, System.Drawing.Imaging.PixelFormat.Format1bppIndexed);

    //save image
    string file = string.Format(Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.Desktop), "test.png"));
    decodedBitmap.Save(file);

    //compress the image again
    byte[] new_z64 = Compress(uzipZ64);
    byte[] newData = new byte[new_z64.Length + 2];
    newData[0] = 120;//e, add the header "eJ"
    newData[1] = 156;//J
    Array.Copy(new_z64, 0, newData, 2, new_z64.Length);
    string newZipString = Convert.ToBase64String(newData); //new result

    //but it's not same from one source.
    bool isSame = zipString.Equals(newZipString);

    Console.WriteLine();
}
```

**Comment by piksel** [comment]
You are comparing the compressed result of some data (no mention of how it was created) with it being recompressed with SharpZipLib. If they are not compressed with the same engine, using the same settings, they most likely will be different.

Created a dotnet fiddle from your example code here:
https://dotnetfiddle.net/yhVsbu

**Comment by ghost** [comment]
HI,
Thanks for your response.

My workflow below:
Step 1: Edit label *.nlbl and print out to the file to .prn in the tool of
Zebra Printer.
Step 2: Because of the fonts not supported in printer, so all the text
converted to images automatically when I print out it to file. (The text is
changeable.)
for example the encoded image with red box.
![image](https://user-images.githubusercontent.com/29701120/201651614-7f82331f-906f-4c76-89b5-0ae53e8986de.png)


step 3:  follow the steps below (The compressed result was exported from
the label, which is created by the tool of Zebra Printer and follow the
below descriptions.):
![image](https://user-images.githubusercontent.com/29701120/201651696-f6c44d9c-de2a-4a2d-b465-7c574aeb5ee4.png)

The operations below were under SharpZipLib. Though the operation in
different engine, the result should be same，if not, each holder has different
private algorithm. The LZ77 is public source code which is not meaningful.
(1) I decompressed the content inside the red box to *.png.
(2) Load the image from file which was saved just now.
(3) However I can not get the content inside the red box after compressed
the image to text again.
The issue relate to LZ77 algorithm,  DeflateStream in C#, zlib.net, SharpZipLib
all can not help me to get a correct result.

Maybe I explained my issue clearly.

nils måsén ***@***.***> 于2022年11月11日周五 17:19写道：

**Reply by piksel** [reply]
Your images was removed from the response. Try adding a response in github directly. 

> each holder has his private algorithm

This is absolutely a possibility, but as the [wikipedia article](https://en.wikipedia.org/wiki/Deflate#Encoder/compressor) states:
> Other Deflate encoders have been produced, all of which will also produce a compatible [bitstream](https://en.wikipedia.org/wiki/Bitstream) capable of being decompressed by any existing Deflate decoder. _**Differing implementations will likely produce variations on the final encoded bit-stream produced**_. The focus with non-zlib versions of an encoder has normally been to produce a more efficiently compressed and smaller encoded stream.

**Reply by ghost** [reply]
HI Sir,  I had updated the images at the previous position. please refer to it. Expect to your response.

**Reply by piksel** [reply]
Yeah, well the second point still stands. DEFLATE:d data is just a set of instructions for how to recreate the original source (INFLATE) and there is no 1:1 mapping from source data to the compressed form. If you need it to be exactly equal, you need to use the **same** library, using the same settings (search window size etc).
